allennlp/training/trainer.py

```diff
@@ -1039,7 +538,8 @@ def _restore_checkpoint(self) -> Tuple[int, List[float]]:
         if batch_num_total is not None:
             self._batch_num_total = batch_num_total
 
-        return epoch_to_return, val_metric_per_epoch
+        return epoch_to_return
+
 
     # Requires custom from_params.
     @classmethod
```

Comment:
Crazy idea, probably a bad idea, but here's a thought (also not super related to this PR):

The reason this needs a custom `from_params` is because there are two constructor arguments that have sequential dependencies with arguments that were constructed previously.  The `Optimizer` needs the `Model`, and the `LearningRateScheduler` needs the `Optimizer` (and this method moves the model and the optimizer to the right cuda device, but that could be done after-the-fact in the `Trainer` constructor without much trouble).  One way to remove the need for a custom `from_params` entirely is to add previously-constructed arguments to the allowed `**extras` dictionary.  So, after you've constructed a `Model`, you can use it for any future arguments that need it.

The reason this is probably a bad idea is that it's quite magical, on top of the already magical `from_params` logic.  But, I don't _think_ it would be too hard, or too hard to document, so maybe it's worth considering?  If other people think this is a good idea, we could open a separate issue for it and fix it some time.